File: src/lib/software/plex_media_server.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class PlexMediaServer(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'plex-media-server'
        self.name = 'Plex Media Server'
        self.category = SoftwareCategory.Media
        self.download_name = 'PlexMediaServer-x86_64.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = False
        self.icon = 'plex-media-server.png'
        self.homepage = 'https://plex.tv'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'https://downloads\.plex\.tv/plex-media-server-new/\d+\.\d+\.\d+\.\d+-[a-f0-9]{9}/windows/PlexMediaServer-\d+\.\d+\.\d+\.\d+-[a-f0-9]{9}-x86_64\.exe')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://plex.tv/api/downloads/5.json'))

File: packages/app/src/services/software/definitions/variants/plexMediaServer.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'plex-media-server';
	public name = 'Plex Media Server';
	public category = [SoftwareCategory.Media];
	public downloadName = 'PlexMediaServer-x86_64.exe';
	public shouldCacheUrl = true;
	public icon = 'plex-media-server.png';
	public homepage = 'https://plex.tv';
	public parent = 'plex';

	public async resolveDownloadUrl() {
		const res = await fetch('https://plex.tv/api/downloads/5.json');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/downloads\.plex\.tv\/plex-media-server-new\/\d+\.\d+\.\d+\.\d+-[a-f0-9]{9}\/windows\/PlexMediaServer-\d+\.\d+\.\d+\.\d+-[a-f0-9]{9}-x86_64\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}
